### PR TITLE
Feature: implement custom qr code shortcode (fixes #8)

### LINKED ISSUE #8: Custom QR code shortcode for ticket template
## Issue Overview

This would allow for including a QR code with custom data (instead of the default QR code that only outputs the REG CODE). This is useful in cases where the event manager wants a QR code for other purposes besides scanning for check-in. 

See also: https://eventespresso.com/wiki/custom-qr-codes-for-tickets/
Note that Google's Chart API is officially end of life. 

## Bug report or feature request?

* [ ] Bug
* [x] Feature
* [ ] Neither

### PR DESCRIPTION
**Pull request by nerrad**
## Problem this Pull Request solves

**NOTE**: This pull will also require some changes to ee core (see [this branch](https://github.com/eventespresso/event-espresso-core/tree/FET/custom-qr-code-requirements) and in particular [this commit](https://github.com/eventespresso/event-espresso-core/commit/5bec305a099048c5c4aa6fbc6d534ea5ebc79e2d))

See #8 for details.  This is a WIP pull that I haven't completed yet.  Currently it:

- Adds the `[CUSTOM_QR_CODE_*]` shortcode php side.
- Modifies the way qr code data is transformed via javascript by embedding it as data properties instead of hidden text in span tags.

I haven't been able to get this to work yet as there are issues with how the main parser library works and nesting issues with shortcodes _inside_ a shortcode (eg. `[CUSTOM_QR_CODE_ content=[RECIPIENT_FNAME]]`).  I'm not sure this is even doable but I did give it a good go while I was working on it.  

## Checklist

* [ ] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.

### CODE REVIEW

File: domain/services/messages/RegisterCustomShortcodes.php [hunk 1]
```diff
@@ -57,6 +65,48 @@ public function registerNewShortcodes($shortcodes, EE_Shortcodes $shortcode_libr
     {
         // shortcodes to add to EE_Ticket_Shortcodes
         if ($shortcode_library instanceof EE_Ticket_Shortcodes) {
+            // add custom_qr shortcode to every possible field and it needs to always be the _last_ shortcode parsed.
+            $shortcodes['[CUSTOM_QRCODE_*]'] = esc_html__(
+                'This is a shortcode used for generating a qrcode for the arbitrary information. Be aware that there are some limits to the amount of information a qr code can contain.  It is recommended that you test the code output.',
```
Comment (tn3rb) [inline domain/services/messages/RegisterCustomShortcodes.php]: > This is a shortcode used for generating a qrcode for the arbitrary information.

this sentence needs to be fixed but I have no idea what it was intended to be.

maybe:

- "This is a shortcode used for generating a qrcode containing arbitrary information."

File: domain/services/messages/RegisterCustomShortcodes.php [hunk 2]
```diff
@@ -57,6 +65,48 @@ public function registerNewShortcodes($shortcodes, EE_Shortcodes $shortcode_libr
...
+            . '</li>'
+            . '<li><strong>d</strong>:'
+            . esc_html__(
+                'You can add a extra param for setting the dimensions of the qr code using "d=20" format.  So [QRCODE_* d=40] will parse to a qrcode that is 40 pixels wide by 40 pixels high.',
+                'event_espresso'
+            )
+            . '</li>'
+            . '<li><strong>color</strong>:'
+            . esc_html__(
+                'Use a hexadecimal color for the qr code color.  So you can do [QRCODE_* color=#f00] to have the code printed in red.',
```
Comment (tn3rb) [inline domain/services/messages/RegisterCustomShortcodes.php]: might be good to link to a site with a list of common color codes ?

**Comment by tn3rb** [comment]
@joshfeck 

I reviewed the work done so far in this PR and the code looks good, but the shortcode attribute descriptions in the admin could use some work.

As well, Darren noted this in the code itself:

```
@todo currently testing this.  It's not working.  I suspect I may need to add a parser_before filter hook in core so that I can add some dummy delimiters.  That way the content won't get picked up by earlier parsers.
I'll also need to rewire the js so that it will pick up multiple instances of qr-code generation nodes.
Likely can just use data attributes instead of the multiple spans in place right now!
```

so this will need to be reassigned to a dev to complete (or you and/or Tony if you want a challenge)

**Comment by nerrad** [comment]
Just a reminder that the original comment for this pull is important as well.  Namely:

- There are/will be changes needed in EE core to support this.  
- There are significant problems with nesting shortcodes within shortcodes (which may prove to be 
- It does NOT work yet.  I need to emphasize this very much was a work in progress and much of the work so far in this pull and in the related core branch was left from my experimentations to _try_ and get this working.

**Pull request closed without merging**